### ISSUE TITLE
Issue with constraints

### ISSUE THREAD

**Issue by emilbayes** [djrobstep/migra#135]
Hi!

I have this SQL schema:

```sql
CREATE SCHEMA auth;

CREATE TABLE auth.users (
  id SERIAL PRIMARY KEY,
  ref UUID,
  active BOOL DEFAULT TRUE,
  info JSONB NOT NULL DEFAULT '{}'::jsonb,
  created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
  updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);

CREATE TABLE auth.provider_local (
  user_id INTEGER UNIQUE REFERENCES auth.users(id),
  username VARCHAR CONSTRAINT username_unique_local UNIQUE,
  password_hash BYTEA,
  created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
  updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);
```

When running `migra --unsafe postgres://a postgres://b` on two databases that have only the above applied, I get the following "diff":

```sql
alter table "auth"."provider_local" drop constraint "provider_local_user_id_fkey";

alter table "auth"."provider_local" drop constraint "provider_local_user_id_key";

alter table "auth"."provider_local" drop constraint "username_unique_local";
```

If I run the following query on the cataloge:

```sql
SELECT con.*
       FROM pg_catalog.pg_constraint con
            INNER JOIN pg_catalog.pg_class rel
                       ON rel.oid = con.conrelid
            INNER JOIN pg_catalog.pg_namespace nsp
                       ON nsp.oid = connamespace
       WHERE nsp.nspname = 'auth'
             AND rel.relname = 'provider_local';
```

I get for `a`:

<table>
	<tr>
		<th>oid</th>
		<th>conname</th>
		<th>connamespace</th>
		<th>contype</th>
		<th>condeferrable</th>
		<th>condeferred</th>
		<th>convalidated</th>
		<th>conrelid</th>
		<th>contypid</th>
		<th>conindid</th>
		<th>conparentid</th>
		<th>confrelid</th>
		<th>confupdtype</th>
		<th>confdeltype</th>
		<th>confmatchtype</th>
		<th>conislocal</th>
		<th>coninhcount</th>
		<th>connoinherit</th>
		<th>conkey</th>
		<th>confkey</th>
		<th>conpfeqop</th>
		<th>conppeqop</th>
		<th>conffeqop</th>
		<th>conexclop</th>
		<th>conbin</th>
	</tr>
	<tr>
		<td>16427</td>
		<td>provider_local_user_id_key</td>
		<td>16392</td>
		<td>u</td>
		<td>0</td>
		<td>0</td>
		<td>1</td>
		<td>16418</td>
		<td>0</td>
		<td>16426</td>
		<td>0</td>
		<td>0</td>
		<td> </td>
		<td> </td>
		<td> </td>
		<td>1</td>
		<td>0</td>
		<td>1</td>
		<td>{1}</td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
	</tr>
	<tr>
		<td>16429</td>
		<td>username_unique_local</td>
		<td>16392</td>
		<td>u</td>
		<td>0</td>
		<td>0</td>
		<td>1</td>
		<td>16418</td>
		<td>0</td>
		<td>16428</td>
		<td>0</td>
		<td>0</td>
		<td> </td>
		<td> </td>
		<td> </td>
		<td>1</td>
		<td>0</td>
		<td>1</td>
		<td>{2}</td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
	</tr>
	<tr>
		<td>16430</td>
		<td>provider_local_user_id_fkey</td>
		<td>16392</td>
		<td>f</td>
		<td>0</td>
		<td>0</td>
		<td>1</td>
		<td>16418</td>
		<td>0</td>
		<td>16406</td>
		<td>0</td>
		<td>16395</td>
		<td>a</td>
		<td>a</td>
		<td>s</td>
		<td>1</td>
		<td>0</td>
		<td>1</td>
		<td>{1}</td>
		<td>{1}</td>
		<td>{96}</td>
		<td>{96}</td>
		<td>{96}</td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
	</tr>
</table>

And for `b`:

<table>
	<tr>
		<th>oid</th>
		<th>conname</th>
		<th>connamespace</th>
		<th>contype</th>
		<th>condeferrable</th>
		<th>condeferred</th>
		<th>convalidated</th>
		<th>conrelid</th>
		<th>contypid</th>
		<th>conindid</th>
		<th>conparentid</th>
		<th>confrelid</th>
		<th>confupdtype</th>
		<th>confdeltype</th>
		<th>confmatchtype</th>
		<th>conislocal</th>
		<th>coninhcount</th>
		<th>connoinherit</th>
		<th>conkey</th>
		<th>confkey</th>
		<th>conpfeqop</th>
		<th>conppeqop</th>
		<th>conffeqop</th>
		<th>conexclop</th>
		<th>conbin</th>
	</tr>
	<tr>
		<td>16445</td>
		<td>provider_local_user_id_key</td>
		<td>16410</td>
		<td>u</td>
		<td>0</td>
		<td>0</td>
		<td>1</td>
		<td>16436</td>
		<td>0</td>
		<td>16444</td>
		<td>0</td>
		<td>0</td>
		<td> </td>
		<td> </td>
		<td> </td>
		<td>1</td>
		<td>0</td>
		<td>1</td>
		<td>{1}</td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
	</tr>
	<tr>
		<td>16447</td>
		<td>username_unique_local</td>
		<td>16410</td>
		<td>u</td>
		<td>0</td>
		<td>0</td>
		<td>1</td>
		<td>16436</td>
		<td>0</td>
		<td>16446</td>
		<td>0</td>
		<td>0</td>
		<td> </td>
		<td> </td>
		<td> </td>
		<td>1</td>
		<td>0</td>
		<td>1</td>
		<td>{2}</td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
	</tr>
	<tr>
		<td>16448</td>
		<td>provider_local_user_id_fkey</td>
		<td>16410</td>
		<td>f</td>
		<td>0</td>
		<td>0</td>
		<td>1</td>
		<td>16436</td>
		<td>0</td>
		<td>16424</td>
		<td>0</td>
		<td>16413</td>
		<td>a</td>
		<td>a</td>
		<td>s</td>
		<td>1</td>
		<td>0</td>
		<td>1</td>
		<td>{1}</td>
		<td>{1}</td>
		<td>{96}</td>
		<td>{96}</td>
		<td>{96}</td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
	</tr>
</table>

So it seems to me that migra is yielding a false positive? Both databases are running Postgres 12.3, but one is running locally and the other is running on Heroku in a hobby tier database

**Comment by emilbayes** [comment]
Oh yeah, and for completeness, here's some system info:

```sh
$ pip --version
pip 20.1.1 from ~/Library/Python/3.7/lib/python/site-packages/pip (python 3.7)

$ pip show migra
Name: migra
Version: 1.0.1593567183
Summary: Like `diff` but for PostgreSQL schemas
Home-page: https://migra.djrobstep.com/
Author: Robert Lechte
Author-email: robertlechte@gmail.com
License: Unlicense
Location: ~/Library/Python/3.7/lib/python/site-packages
Requires: schemainspect, six, sqlbag
Required-by:

$ pip show schemainspect
Name: schemainspect
Version: 0.1.1593605722
Summary: Schema inspection for PostgreSQL (and possibly others)
Home-page: https://github.com/djrobstep/schemainspect
Author: Robert Lechte
Author-email: robertlechte@gmail.com
License: Unlicense
Location: ~/Library/Python/3.7/lib/python/site-packages
Requires: sqlalchemy, six
Required-by: migra

$ pip show sqlalchemy
Name: SQLAlchemy
Version: 1.3.18
Summary: Database Abstraction Library
Home-page: http://www.sqlalchemy.org
Author: Mike Bayer
Author-email: mike_mp@zzzcomputing.com
License: MIT
Location: ~/Library/Python/3.7/lib/python/site-packages
Requires: 
Required-by: sqlbag, schemainspect

$ pip show sqlbag
Name: sqlbag
Version: 0.1.1579049654
Summary: various snippets of SQL-related boilerplate
Home-page: https://github.com/djrobstep/sqlbag
Author: Robert Lechte
Author-email: robertlechte@gmail.com
License: UNKNOWN
Location: ~/Library/Python/3.7/lib/python/site-packages
Requires: pathlib, six, sqlalchemy
Required-by: migra
```

**Comment by emilbayes** [comment]
If I run the command for constraints I dug out from schemainspect I get:

`a`:

<table>
	<tr>
		<th>schema</th>
		<th>name</th>
		<th>table_name</th>
		<th>definition</th>
		<th>constraint_type</th>
		<th>index</th>
		<th>extension_oid</th>
	</tr>
	<tr>
		<td>auth</td>
		<td>provider_local_user_id_fkey</td>
		<td>provider_local</td>
		<td>FOREIGN KEY (user_id) REFERENCES auth.users(id)</td>
		<td>FOREIGN KEY</td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
	</tr>
	<tr>
		<td>auth</td>
		<td>provider_local_user_id_key</td>
		<td>provider_local</td>
		<td>UNIQUE (user_id)</td>
		<td>UNIQUE</td>
		<td>provider_local_user_id_key</td>
		<td><i>NULL</i></td>
	</tr>
	<tr>
		<td>auth</td>
		<td>username_unique_local</td>
		<td>provider_local</td>
		<td>UNIQUE (username)</td>
		<td>UNIQUE</td>
		<td>username_unique_local</td>
		<td><i>NULL</i></td>
	</tr>
	<tr>
		<td>auth</td>
		<td>users_pkey</td>
		<td>users</td>
		<td>PRIMARY KEY (id)</td>
		<td>PRIMARY KEY</td>
		<td>users_pkey</td>
		<td><i>NULL</i></td>
	</tr>
</table>


`b`:

<table>
	<tr>
		<th>schema</th>
		<th>name</th>
		<th>table_name</th>
		<th>definition</th>
		<th>constraint_type</th>
		<th>index</th>
		<th>extension_oid</th>
	</tr>
	<tr>
		<td>auth</td>
		<td>provider_local_user_id_fkey</td>
		<td>provider_local</td>
		<td>FOREIGN KEY (user_id) REFERENCES auth.users(id)</td>
		<td>FOREIGN KEY</td>
		<td><i>NULL</i></td>
		<td><i>NULL</i></td>
	</tr>
	<tr>
		<td>auth</td>
		<td>provider_local_user_id_key</td>
		<td>provider_local</td>
		<td>UNIQUE (user_id)</td>
		<td>UNIQUE</td>
		<td>provider_local_user_id_key</td>
		<td><i>NULL</i></td>
	</tr>
	<tr>
		<td>auth</td>
		<td>username_unique_local</td>
		<td>provider_local</td>
		<td>UNIQUE (username)</td>
		<td>UNIQUE</td>
		<td>username_unique_local</td>
		<td><i>NULL</i></td>
	</tr>
	<tr>
		<td>auth</td>
		<td>users_pkey</td>
		<td>users</td>
		<td>PRIMARY KEY (id)</td>
		<td>PRIMARY KEY</td>
		<td>users_pkey</td>
		<td><i>NULL</i></td>
	</tr>
</table>

**Comment by djrobstep** [comment]
Hey, I can't seem to reproduce this with the latest migra/schemainspect. I get no differences, as expected.

What version of postgres are you running?

**Comment by emilbayes** [comment]
I’m on 12.3 locally and 12.2 remote. But I do also get a blank diff with the above, sorry! I tried to leave out all the irrelevant parts. I’ll rework my minimal example and come back

**Comment by emilbayes** [comment]
@djrobstep Apologies, it turns out the "bug" is in the user used to connect to the remote had only Heroku's definition of a "Read-only" credential. I was not able to find documentation on exactly, but it seems that it only includes "SELECT" on the table (checked `information_schema.role_*_grants`)

**Comment by djrobstep** [comment]
No worries. Which information was it unable to read? File a separate issue if it didn't handle the lack of permissions as well as you'd like (it should tell you if it can't read the information it needs).
